src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+
+                <!--
+                  Override targets which don't work with project files that are not present on disk.
+                  See https://github.com/NuGet/Home/issues/14148.
+                -->
+
+                <Target Name="_FilterRestoreGraphProjectInputItems"
+                        DependsOnTargets="_LoadRestoreGraphEntryPoints"
+                        Returns="@(FilteredRestoreGraphProjectInputItems)">
+                  <ItemGroup>
+                    <FilteredRestoreGraphProjectInputItems Include="@(RestoreGraphProjectInputItems)" />
+                  </ItemGroup>
+                </Target>
+
+                <Target Name="_GetAllRestoreProjectPathItems"
+                        DependsOnTargets="_FilterRestoreGraphProjectInputItems"
+                        Returns="@(_RestoreProjectPathItems)">
+                  <ItemGroup>
+                    <_RestoreProjectPathItems Include="@(FilteredRestoreGraphProjectInputItems)" />
+                  </ItemGroup>
+                </Target>
+
+                <Target Name="_GenerateRestoreGraph"
+                        DependsOnTargets="_FilterRestoreGraphProjectInputItems;_GetAllRestoreProjectPathItems;_GenerateRestoreGraphProjectEntry;_GenerateProjectRestoreGraph"
+                        Returns="@(_RestoreGraphEntry)">
+                  <!-- Output from dependency _GenerateRestoreGraphProjectEntry and _GenerateProjectRestoreGraph -->
+                </Target>
+            </Project>
+            """;
+        // <EnableDefaultItems>false</EnableDefaultItems>
+        // <Compile Include="{documentFilePath}" />
+    }
+
+    /// <summary>
+    /// Removes a document with the matching file path from this workspace.
+    /// 
+    /// Calls to this method and <see cref="AddMiscellaneousDocument(Uri, SourceText, string, ILspLogger)"/> are made
+    /// from LSP text sync request handling which do not run concurrently.
+    /// </summary>
+    public void TryRemoveMiscellaneousDocument(Uri uri, bool removeFromMetadataWorkspace)
+    {
+        if (removeFromMetadataWorkspace)
+        {
+            return;
+        }
+
+        // We'll only ever have a single document matching this URI in the misc solution.
+        var matchingDocument = CurrentSolution.GetDocumentIds(uri).SingleOrDefault();
+        if (matchingDocument != null)
+        {
+            if (CurrentSolution.ContainsDocument(matchingDocument))
+            {
+                OnDocumentRemoved(matchingDocument);
+            }
+            else if (CurrentSolution.ContainsAdditionalDocument(matchingDocument))
+            {
+                OnAdditionalDocumentRemoved(matchingDocument);
+            }
+
+            // Also remove the project - we always create a new project for each misc file we add
+            // so it should never have other documents in it.
+            var project = CurrentSolution.GetRequiredProject(matchingDocument.ProjectId);
+            OnProjectRemoved(project.Id);
```

Comment:
As you figure out how to reuse code, you'll have to deal with the race here that you may be removing a project that's still being updated elsewhere.